### PR TITLE
Support for use of Enums in MultiIndex

### LINKED ISSUE #21298: Support for Enums in MultiIndex
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
from enum import Enum

MyEnum = Enum("MyEnum", "A B")

df = pd.DataFrame(columns=pd.MultiIndex.from_product(iterables=[MyEnum, [1, 2]]))  # TypeError: 'values' is not ordered, please explicitly specify the categories order by passing in a categories argument.

df = pd.DataFrame(columns=pd.MultiIndex.from_product(iterables=[pd.Series(MyEnum, dtype="category"), [1, 2]]))  # this workaround successfully executes, but...
df.append({(MyEnum.A, 1): "abc", (MyEnum.B, 2): "xyz"}, ignore_index=True)  # ... this "append" statement then raises the same error.

df.loc[0, [(MyEnum.A, 1), (MyEnum.B, 2)]] = 'abc', 'xyz'  # this works, but is less desirable (can't pass a dict, need to come up with a row indexer, etc.)

```
#### Problem description

Though Enums can easily be used as column indexers, strange errors appear to arise when they are used (as one of the factors) in a MultiIndex.

The multiindex (and dataframe) can be created successfully if an (ordered) categorical Series is passed to the constructor. Yet in this case, appending rows in the usual way fails. One can create new rows using `.loc`, and yet this is not as nice.

This whole situation can be avoided by using strings instead of an Enum. Alternatively, one can use an IntEnum---and yet this essentially uses the underlying integers, instead of the names, as the column indexers.

As the use of enums as columns is perfectly supported in the case of a simple index, it seems a shortcoming that they can't be used in a MultiIndex.

#### Expected Output
```python
>>> df = pd.DataFrame(columns=pd.MultiIndex.from_product(iterables=[MyEnum, [1, 2]]))
>>> df
Empty DataFrame
Columns: [(MyEnum.A, 1), (MyEnum.A, 2), (MyEnum.B, 1), (MyEnum.B, 2)]
Index: []
>>> df.append({(MyEnum.A, 1): "abc", (MyEnum.B, 2): "xyz"}, ignore_index=True)
  MyEnum.A      MyEnum.B     
         1    2        1    2
0      abc  NaN      NaN  xyz
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Darwin
OS-release: 17.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.0
pytest: None
pip: 10.0.1
setuptools: 39.0.1
Cython: None
numpy: 1.13.3
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by benediamond**
- [x] closes #21298
- [x] tests added / passed: added `test_use_enum_in_multiindex` in `tests/indexes/test_multi.py`, see diff
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The boolean flag in https://github.com/pandas-dev/pandas/blob/9f95f7dbffef7752175ca9ed918314cb6f0b9b18/pandas/core/arrays/categorical.py#L2526-L2528 incorrectly caused a `TypeError` to be raised in
https://github.com/pandas-dev/pandas/blob/9f95f7dbffef7752175ca9ed918314cb6f0b9b18/pandas/core/arrays/categorical.py#L352-L354
when an Enum was used in a MultiIndex. No tests fail when the flag is flipped back. This is fixed in this PR.

### CODE REVIEW

**Comment by benediamond** [comment]
failure is due to unavailability of `enum` in test environment...

**Comment by gfyoung** [comment]
> CircleCI failed due to unavailability of module enum in test environment. I don't know how else to test this---should I simply remove the new test I added?

File: pandas/tests/indexes/test_multi.py
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: what is all this for?

simply construct the expected index directly

File: pandas/tests/indexes/test_multi.py
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: all imports go at the top

File: pandas/tests/indexes/test_multi.py
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: you would have to add ``enum34`` (the 2.7 backport) in our 2.7 builds

```
(pandas) bash-3.2$ grep python=2.7 ci/*.yaml
appveyor-27.yaml:  - python=2.7.*
circle-27-compat.yaml:  - python=2.7*
travis-27-locale.yaml:  - python=2.7
travis-27.yaml:  - python=2.7*
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/test_multi.py
Comment (benediamond) [reply pandas/tests/indexes/test_multi.py]: The thought here is that the expected index _cannot_ be constructed directly, due to the exact bug I'm trying to fix. Thus instead I create an expected "level", and verify that it matches the relevant level of `df`'s index. This is done by replicating a sub-step of `tm.assert_index_equal`.

File: pandas/tests/indexes/test_multi.py
Comment (benediamond) [reply pandas/tests/indexes/test_multi.py]: I'm sorry, does this mean simply importing `enum34` inline? Thank you.

Comment (benediamond) [reply pandas/tests/indexes/test_multi.py]: ok, I've hopefully done this correctly.

**Comment by benediamond** [comment]
@jreback any idea why it's marking this `whatsnew` line as a conflict...?

**Comment by benediamond** [comment]
@gfyoung I'm stumped by this `whatsnew` conflict, any thoughts?

**Comment by gfyoung** [comment]
> @gfyoung I'm stumped by this whatsnew conflict, any thoughts?

@benediamond : I would actually move your `whatsnew` addition to `0.24.0`.  That should take care of that 😄

File: doc/source/whatsnew/v0.23.1.txt
Comment (gfyoung) [inline doc/source/whatsnew/v0.23.1.txt]: > @gfyoung I'm stumped by this whatsnew conflict, any thoughts?

@benediamond : I would actually move your `whatsnew` addition to `0.24.0`.   That will suffice. 😄

Comment (benediamond) [reply doc/source/whatsnew/v0.23.1.txt]: I've gone ahead and just removed it---I'm not sure where it belongs, as it appeared that this PR (when it was passing) was slated for the next milestone. It was: "Bug in :func:`pandas.core.arrays.categorical._factorize_from_iterable` inappropriately caused ``TypeError`` to be raised when an ``Enum`` was used as a factor in a ``MultiIndex`` (:issue:`21298`)"

Comment (gfyoung) [reply doc/source/whatsnew/v0.23.1.txt]: @benediamond : I would put it in `0.24.0`, as I suggested above.  We will need a `whatsnew` for this for sure, and that would be the best place to put it.

Comment (benediamond) [reply doc/source/whatsnew/v0.23.1.txt]: @gfyoung Done. this one didn't create merge conflicts, thankfully

File: pandas/tests/indexes/test_multi.py
Comment (gfyoung) [inline pandas/tests/indexes/test_multi.py]: Given that you use this function only once, I think this abstraction is unnecessary as @jreback alluded to.

File: pandas/tests/indexes/test_multi.py
Comment (gfyoung) [inline pandas/tests/indexes/test_multi.py]: By "top", @jreback meant "top of the file" 😄

Comment (benediamond) [reply pandas/tests/indexes/test_multi.py]: Hah whoops. done

File: pandas/tests/indexes/test_multi.py
Comment (benediamond) [reply pandas/tests/indexes/test_multi.py]: Ok, amended. Thanks for your guidance.

File: pandas/core/arrays/categorical.py
Comment (gfyoung) [inline pandas/core/arrays/categorical.py]: Interesting that this didn't cause any testing issues...

Comment (benediamond) [reply pandas/core/arrays/categorical.py]: Yes, it's striking. It's possible that that`multi.py` is essentially the only caller of `categorical._factorize_from_iterable`, and that the calls that `multi` makes happen to be such that within `Categorical`'s constructor, control is always routed so that the only time that `ordered` is used is to incorrectly raise the `TypeError` on lines 352-354. That's my best-guess explanation, at least.

**Review by gfyoung — APPROVED** [review]
Well, this looks pretty good to me!

cc @jreback

**Comment by benediamond** [comment]
@jreback bumping this.

File: ci/circle-27-compat.yaml
Comment (jreback) [inline ci/circle-27-compat.yaml]: can you remove from all but 1 of the 27 compat tests (the travis-27) one

File: pandas/tests/indexes/test_multi.py
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: can you use ``@td.skip_if_no('enum')`` here (then do the actual import inside), we have to guard against the tests being run on 2.7 w/o the compat library (by a user).

File: pandas/tests/indexes/test_multi.py
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: this is pretty complicated, why cannot you just directly construct the Index here?

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew note for 0.24

**Comment by benediamond** [comment]
for whatsnew, see: https://github.com/benediamond/pandas/blob/4cd2dff96775a828687acd7c65919ec135824cf4/doc/source/whatsnew/v0.24.0.txt#L132-L135

**Comment by benediamond** [comment]
> can you use @td.skip_if_no('enum') here (then do the actual import inside), we have to guard against the tests being run on 2.7 w/o the compat library (by a user).

@jreback implemented here:
```python
    @td.skip_if_no('enum')
    def test_use_enum_in_multiindex(self):
        # GH 21298
        # Allow use of Enums as one of the factors in a MultiIndex.
        from enum import Enum
        MyEnum = Enum("MyEnum", "A B")
        ...
```
but yields:
```
==================================== ERRORS ====================================
_____________ ERROR collecting pandas/tests/indexes/test_multi.py ______________
pandas/tests/indexes/test_multi.py:30: in <module>
    class TestMultiIndex(Base):
pandas/tests/indexes/test_multi.py:3310: in TestMultiIndex
    @td.skip_if_no('enum')
E   NameError: name 'td' is not defined
------- generated xml file: /tmp/circle-junit.2ZpyFLd/reports/junit.xml --------
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
===================== 3 skipped, 1 error in 43.23 seconds ======================
```
not sure what's going on 😩

**Comment by TomAugspurger** [comment]
I think you need to `import pandas.util._test_decorators as td` at the top of `test_multi.py`

**Comment by jreback** [comment]
closing as stale, if you'd like to continue, pls ping.

**Comment by benediamond** [comment]
@jreback, my understanding was that this commit was ready, but had been waiting for your approval. I found it hard to get responses from you. if you're ready to go, i'm ready to play my part. it should be almost there.

File: doc/source/whatsnew/v0.24.0.txt
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.txt]: Release notes shouldn't reference private methods.

You should be able to reference Enum with :class:`enum.Enum`, and :class:`MultiIndex`.

**Comment by benediamond** [comment]
@jreback and others: looks like this issue has _now been fixed_ by PR #22072, see also issue #15457. I have gone ahead and preserved the test I wrote in my rebase, in case you'd like to merge it in---if not no worries and this (as well as issue #21298) can be closed! (note that issue #15457 was a somewhat different problem). thanks for everyone's time and patience.

**Comment by WillAyd** [comment]
@benediamond closing per your latest comment. Ping if we need to reopen

**Comment by scratchmex** [comment]
I want to point out that when merging dataframes with enums as multi-index it throws
`RuntimeWarning: The values in the array are unorderable. Pass `sort=False` to suppress this warning.`
so I think the support for enums is not yet completed

**Comment by jreback** [comment]
you could open a separate issue with a reproducible example

but we have no official support for enums and would need a champion in the community

**Pull request closed without merging**
